framework/src/main/java/org/checkerframework/framework/util/typeinference8/constraint/TypeConstraint.java

```diff
@@ -27,14 +28,65 @@ public abstract class TypeConstraint implements Constraint {
   /** T, the type on the right hand side of the constraint; may contain inference variables. */
   protected AbstractType T;
 
+  /** The constraint whose reduction created this constraint. */
```

Comment:
Please comment on the relationship between these two variables.  Is one of them always null?  Is exactly one of them always null?  (I think the latter is not true because both can be null, but I'm not sure about that.)